### DISCUSSION TITLE
Best way to secure /admin routes

### DISCUSSION THREAD

**Discussion by GriffinSauce** [vercel/next.js#10724] [Help]
Hi, so I'm slowly getting my head around sessions and auth with Next.js (deployed on Now). What I'm running into now is that I have a HOC `ensureAdmin` that I can wrap a page with to check that a user is authenticated and actually an admin. (the API routes are secured ofc so it's mostly a UX/visibility thing). The downside is that you need to repeat this for every page.

Is there a way to do this globally for everything under this path /admin? It would be nice to for instance have a custom `_app.js` that applies to this folder+children or something.

How do you solve this?
Should I just not try to limit access to the page and use a fetch hook to check client side after rendering the skeleton?

**Comment by janus-reith** [comment]
At which point do you have access to the authentication state?
You could create an api route with the catchall segment, e.g. `api/authRoutes/[..slug].js`

Then you could use the new redirect feature like this:

```
experimental: {
    redirects() {
      return [
        {
          source: "/admin/:authRoute*",
          destination: "/api/checkAuth/admin/:authRoute*",
          permanent: true
        }
      ];
    }
  }
```

`checkAuth` would be an API Endpoint with `res` and `req` and you could use headers/cookies/whatever from the request there to determine if authenticated and redirect to the actual route based on that.
In that function you could do something like this `const { query: { slug } } = req` and `slug` would then be an array of all sub-paths, in this case ["admin", "yourRouteName"] which you could join again to get the original route the client requested and redirect to it.

This way you have no added loading time and don't need to redirect clientside.

**Reply by GriffinSauce** [reply]
Thanks for the insights! That's a cool solution.

I set a token in a cookie but no sessions otherwise. So I have to still also grab the user from db and check their permissions. 

> This way you have no added loading time and don't need to redirect clientside.

This would be because the Now proxy handles redirects and you're not rendering pages to do them, correct? Just making sure I fully understand.

**Reply by timneutkens** [reply]
Creating a redirect to a serverless function is not the best idea, it'd be the same as server-rendering the page.

> This way you have no added loading time and don't need to redirect clientside.

This is incorrect, you're redirecting to an API route, API routes are serverless functions, so you do add load time. On top of that the result can't be cached on a CDN so it always has to go to the origin region (which could be on the other side of the world).


The way that we built the zeit.co dashboard (fully static):
- Uses https://github.com/zeit/swr to ensure data is up-to-date while staying fast
- All pages are auto-exported to static HTML, no `getInitialProps` on any pages
- All pages implement a loading shell, an example of this is: https://zeit.co/dashboard/abcdefgh
- When a user opens the page a fetch is done against the API, if you're logged in they have a cookie set already, if not you're redirected to the login screen

**Reply by janus-reith** [reply]
Oh, sorry If I got that part wrong.
So each time my api route is invoked, it completely "boots" from scratch or where does it add time? I get the CDN part, but apart from that I thought this is still way better than a client redirect or using initalprops for the whole page. I wonder how to handle this logic else?

> When a user opens the page a fetch is done against the API, if you're logged in they have a cookie set already, if not you're redirected to the login screen

Could to elaborate further on that? I mean, basically thats the same functionality I would expect the described API route to do. Where does the part sit that does this fetch and redirects? Outside of nextjs at all as some kind of router?
The little ui flicker on https://zeit.co/dashboard/abcdefgh suggests that this is happening clientside?

Also, are these considerations only necesary for serverless deployments or in general with api routes?

**Reply by timneutkens** [reply]
> So each time my api route is invoked, it completely "boots" from scratch or where does it add time? 

Not each time, serverless functions boot up and wind down though.

> I get the CDN part, but apart from that I thought this is still way better than a client redirect or using initialprops for the whole page. I wonder how to handle this logic else?

Client-side handling allows for preloading the API url and a bunch of other optimizations like showing a spinner while data is being loaded.

> Could to elaborate further on that? I mean, basically thats the same functionality I would expect the described API route to do. Where does the part sit that does this fetch and redirects? Outside of nextjs at all as some kind of router?

We have a custom hook that looks something like this (removed a bunch of business logic to make it easier to understand):

```js
import { useEffect } from 'react'
import Router from 'next/router'
import useSWR from 'swr'

import { getUser } from '@lib/authenticate'
import { API_USER } from '@lib/api-endpoints'

export default function useUser() {
  const swr = useSWR(API_USER, () => getUser())

  useEffect(() => {
    if (swr.data && !swr.data.user && redirect) {
      // redirect to login
      Router.replace(
        `/login`
      )
    }
  }, [swr.data])

  swr.isLoggedOut = swr.data && !swr.data.user
  swr.user = swr.data?.user

  return swr
}
```

It just uses basic Next.js features + SWR basically

> Also, are these considerations only necesary for serverless deployments or in general with api routes?

When building dashboards we found the approach we use for zeit.co/dashboard to hold the best performance and scaling characteristics, so I'd definitely use this approach in all cases of deployment.

**Reply by janus-reith** [reply]
Ah, so basically the most simple approach like I also often did with some Apollo fetching, use an Effect, make sure data finsihed loading, if finished check if present, else redirect.

So agreed, OP asked for the best way for a similar usecase like the one you described and you seem to have more experience with the actual usage in production, so that makes sense.
(Still surprised though).

But out of curiosity, how would such an approach work when not using something like a dashboard, but truly "secret" data? Like, lets say I have a range of secured pages with data only available to a range of authenticated users, and still want them to be fully statically prerendered for performance reasons. It seems like using the API routes then would still be the only out of the box solution besides using initialprops?
(To give context, I could think of product pages including non-public data like prices, and those product pages are only available to certain logged in users, which while logged in might still get benefits from havig a static page, e.g on a weak mobile connection).

And also regarding that, your comparison of API routes with  server-rendering the page made me wonder. Just be safe I got things right, I would assume using an API route and redirecting to a satatic page then would still involve less computing time than using initialprops and running the whole render on the server on demand?

**Reply by GriffinSauce** [reply]
Thanks for the input Tim. The part about "secret" data approaches my use case. Basically I want to limit access to the page altogether so that we can be "dumb" about it beyond that point.

Perhaps that's not a great strategy, using a hook like this definitely looks more ergonomic. (and allows static hosting). It's a bit of a paradigm flip from "limit access to all these express routes by chucking in some middleware".

**Reply by timneutkens** [reply]
> But out of curiosity, how would such an approach work when not using something like a dashboard, but truly "secret" data? Like, lets say I have a range of secured pages with data only available to a range of authenticated users, and still want them to be fully statically prerendered for performance reasons. It seems like using the API routes then would still be the only out of the box solution besides using initialprops?
(To give context, I could think of product pages including non-public data like prices, and those product pages are only available to certain logged in users, which while logged in might still get benefits from havig a static page, e.g on a weak mobile connection).

How would that be different from the dashboard case? You would still do the fetching client-side and serve the static content. It then updates client-side when you're logged in.

> And also regarding that, your comparison of API routes with server-rendering the page made me wonder. Just be safe I got things right, I would assume using an API route and redirecting to a satatic page then would still involve less computing time than using initialprops and running the whole render on the server on demand?

It would be slightly faster than rendering the full page on-demand, but in the case that you described you would end up redirecting to a static page, and that static page still has to fetch the data for that user, so it's slower, harder to scale and less cost-effective. Because serving the static file is significantly faster (with a CDN), scales to more users and you're not paying for server execution).

> Thanks for the input Tim. The part about "secret" data approaches my use case. Basically I want to limit access to the page altogether so that we can be "dumb" about it beyond that point.

> Perhaps that's not a great strategy, using a hook like this definitely looks more ergonomic. (and allows static hosting). It's a bit of a paradigm flip from "limit access to all these express routes by chucking in some middleware".

Data would always come from the API and be behind authentication so this is not really a concern, if you don't want to render a shell you could also return a simple spinner for example. For zeit.co/dashboard we went with the very fancy approach of having the whole layout visible while loading but this is not required.

**Reply by janus-reith** [reply]
@timneutkens No, in my described usecase that product page would be static and not contain data specific per user.
So I wouldn't do the data fetching client side, the product page relies on just the product data, the user is only relevant here during the redirect to determine if that page can be accessed or not.

The product api is restricted with an authorization header, but in this case I would need to make 
sure that nextjs has some authentication present during static generation to fetch that data. But as soon as that page was statically prerendered, anyone accessing it would get the data there without being authorized.

In a real usecase there would probably be some user data fetched on the client aswell, but afterwards, when the relevant page with product data is there already, thereby preventing a loading spinner from showing initally until the product data query are resolved.

Sure could save some effort here and just put the product data fetching client side and would practically also just resort to that in most cases, even more that SEO is not relevant for these authenticated pages, but it's about the principle: Some authorized user on a slow connection could still have a better experience this way.

I guess there might also be other such edge cases besides my product example.
BTW, hope I'm not hijacking this discussion too much, but I'm really interested on the theoretical aspects of this ;)

**Reply by timneutkens** [reply]
> No, in my described usecase that product page would be static and not contain data specific per user.
So I wouldn't do the data fetching client side, the product page relies on just the product data, the user is only relevant here during the redirect to determine if that page can be accessed or not.

First up I'm not exactly understanding your last message but I did sort of understand this first bit.

Are you saying "What if I have a static page that requires authentication" and can't "leak" the actual data that will be shown when a user is logged in?

If it requires authentication it always requires user data, because the question "Is the user logged in" is user data.

As such what you would do is still render a spinner or skeleton.

I realized I hadn't covered why exactly rendering a spinner from CDN is faster than a server-rendered response / API route. For zeit.co/dashboard we did a lot of investigation into making interaction faster, specifically preloading fetch requests before they actually happen (eg using the `<link rel="preload">` tag). In general React hydration takes a bit of time (depending on app size / complexity), if you send down the shell (spinner or skeleton) early (from CDN) React will boot up while the data that is going to be fetched is preloaded, resulting in a faster time to interactive.

**Reply by timneutkens** [reply]
If you're saying "This page is public but I still want to show a user state" (eg user avatar) we handle that on zeit.co client-side as well, eg if you're logged in and go to zeit.co/blog you'll see your avatar in the top right.

**Reply by janus-reith** [reply]
@timneutkens The user data part is clear for me so far, I also did some implementations using nextjs  where user fetching would work both for SSR and Static requests either so I think I get the basic concepts so far.
Sorry if the main point I'm trying to discuss here is so hard to understand, will try to rephrase it:

Regarding the main question:
> Are you saying "What if I have a static page that requires authentication" and can't "leak" the actual data that will be shown when a user is logged in?

Yes, that. Or put into other words: If I need to validate params from the incoming request to determine page access based on that, what is my best option besides using API routes(or initialprops), or is there even any that I could use with nextjs out of the box?

> If it requires authentication it always requires user data, because the question "Is the user logged in" is user data.

Sure, wouldn't argue here. But what I'm talking about is, where said user data is used in any way to determine access to aforementioned page.
If my API requires authentication to return the data necessary to build that static page, this authentication could be anything generic I pass to nextjs for the static build, and not necessarily that of any of the users visiting that page later.

Instead of having the client fetch the data on that page while authenticating that call with the cookie/token/header/, and thereby needing to have that first-time query for data happening on the client, I might want to just have that page available statically, with all necessary but potentially sensitive data included. The server would parse necessary auth data from the request, and decide if that user is allowed to view the page.
To have a little avatar and username at the top right, I certainly still do a separate request for user data on my client based on the very same cookie or whatever mechanism, but that part should not matter for the available data.

**Reply by timneutkens** [reply]
> Instead of having the client fetch the data on that page while authenticating that call with the cookie/token/header/, and thereby needing to have that first-time query for data happening on the client, I might want to just have that page available statically, with all necessary but potentially sensitive data included. The server would parse necessary auth data from the request, and decide if that user is allowed to view the page.

This would always mean the static page is accessible by knowing the path, meaning it can be guessed and accessed. API routes can't prevent access to static pages, only redirect to them, otherwise the page wouldn't be static.

**Reply by janus-reith** [reply]
@timneutkens, oh ok, this really was a missing piece for me. I get that this is the case for static/serverless deployments, but didn't think that it would always be this way.

**Reply by trusheim** [reply]
> When building dashboards we found the approach we use for zeit.co/dashboard to hold the best performance and scaling characteristics, so I'd definitely use this approach in all cases of deployment.

I just wanted to add a +1 to this discussion: it would be great to add some sort of support for static pre-rendering mixed with SSR. My use-case is similar to those in this thread: I'm generating an internal dashboard. Access is controlled via server-side authentication. However, the data in the internal dashboards are expensive to generate (20-30s). I'd prefer to pre-render all of the dashboard pages (e.g. hourly via a cron job, etc) and have server-side control over, basically, whether they're served.

The second-best solution (currently implemented) is to have the recurring process cache results and use SWR to pull down the cached results, but that's one more tool to manage, vs. just calling the build step hourly via deploy hooks.

**Reply by cdock1029** [reply]
@timneutkens Hi Tim, on Full Stack Radio podcast you described how with `getServerSideProps` being able to directly access database because it is excluded at build, this opened up clean usage of backend data access without needing a middle api layer, and that you guys might leverage this on vercel website.

Any heuristics as to when you think this is a better choice than loading skeleton shell and fetching data client side.. or what use cases your team is choosing `getServerSideProps` over swr?

**Reply by smblee** [reply]
Bumping an old thread but...

> > So each time my api route is invoked, it completely "boots" from scratch or where does it add time?
> 
> Not each time, serverless functions boot up and wind down though.
> 
> > I get the CDN part, but apart from that I thought this is still way better than a client redirect or using initialprops for the whole page. I wonder how to handle this logic else?
> 
> Client-side handling allows for preloading the API url and a bunch of other optimizations like showing a spinner while data is being loaded.
> 
> > Could to elaborate further on that? I mean, basically thats the same functionality I would expect the described API route to do. Where does the part sit that does this fetch and redirects? Outside of nextjs at all as some kind of router?
> 
> We have a custom hook that looks something like this (removed a bunch of business logic to make it easier to understand):
> 
> ```js
> import { useEffect } from 'react'
> import Router from 'next/router'
> import useSWR from 'swr'
> 
> import { getUser } from '@lib/authenticate'
> import { API_USER } from '@lib/api-endpoints'
> 
> export default function useUser() {
>   const swr = useSWR(API_USER, () => getUser())
> 
>   useEffect(() => {
>     if (swr.data && !swr.data.user && redirect) {
>       // redirect to login
>       Router.replace(
>         `/login`
>       )
>     }
>   }, [swr.data])
> 
>   swr.isLoggedOut = swr.data && !swr.data.user
>   swr.user = swr.data?.user
> 
>   return swr
> }
> ```
> 
> It just uses basic Next.js features + SWR basically
> 
> > Also, are these considerations only necesary for serverless deployments or in general with api routes?
> 
> When building dashboards we found the approach we use for zeit.co/dashboard to hold the best performance and scaling characteristics, so I'd definitely use this approach in all cases of deployment.


How would you scale this for sub-routes? (e.g. if all routes under `/dashboard` (/dashboard/*) need to be protected).

**Comment by vvo** [comment]
Hey there, following the best practices from Zeit in @timneutkens comment here: https://github.com/zeit/next.js/discussions/10724#discussioncomment-726

I created an example that follows those practices, see it here: https://github.com/zeit/next.js/tree/canary/examples/with-iron-session

let me know what you think

**Reply by GriffinSauce** [reply]
Wow that's awesome! Will give it a look soon, I currently have a different solution for sessions but this looks a lot cleaner.

**Reply by sandys** [reply]
@vvo have you seen this ? https://github.com/zeit/next.js/tree/canary/examples/api-routes-apollo-server-and-client-auth
how is your solution different ?

**Reply by vvo** [reply]
@sandys I am not sure what are the similarities between with-iron-session and the apollo server example? Mine is not based on GraphQL and I cannot see any useUser inside api-routes-apollo-server-and-client-auth

**Reply by sandys** [reply]
@vvo oh - wasnt focusing on the graphql part. that part is incidental. im kind of wondering if the HOC and the cookie behavior there is equivalent to your iron-session ?

e.g. https://github.com/zeit/next.js/blob/canary/examples/api-routes-apollo-server-and-client-auth/apollo/resolvers.js and https://github.com/zeit/next.js/blob/canary/examples/api-routes-apollo-server-and-client-auth/apollo/client.js

**Reply by vvo** [reply]
can't tell exactly, apollo example is using JWT, I am using https://github.com/vvo/next-iron-session and https://github.com/vvo/iron-store behind the scenes. JWT is signed but not encrypted, I think we are diverging from the original question here though :D

**Reply by felixmosh** [reply]
@vvo Thank you for sharing this example.
There is one issue with the fact that the Layout component is part of the page, it get's re-render every page navigation.

**Reply by vvo** [reply]
@felixmosh indeed, since then I have been using techniques of "persistent layout" as shown here: https://adamwathan.me/2019/10/17/persistent-layout-patterns-in-nextjs/. I will update next-iron-session with it.

**Reply by lww** [reply]
Hey @vvo, thanks for the example.
How would you make additional API requests on the profile page only if user is logged in?
I want to prevent sending the requests, if there is no user.

**Reply by vvo** [reply]
Hey there, the best way to do it is to redirect people when they try to access pages that requires to be logged in.

You can find it in the example here: https://github.com/vvo/next-iron-session/blob/25032de07dc932396337427d43c9fc9a316b897c/examples/next.js/pages/profile-sg.jsx#L6-L10

Open an issue in next-iron-session if you need more information

**Comment by sandys** [comment]
@timneutkens how would you do oauth in this situation ? i saw the solution you posted with SWR, but im now confused how would oauth based authentication work in this case - e.g. google auth kind of thing. 

others have asked this question in other places, https://github.com/IdentityModel/oidc-client-js/issues/809

My whole site is SSR.

**Reply by balazsorban44** [reply]
I can highly recommend https://github.com/auth0/nextjs-auth0

We tried oidc-client, and it was a hassle. Bad documentation, and for me at least hard to read source code. Auth0's package is oidc compliant, workes like charm for us - both for api routes and client side stuff - with our custom identity provider (so not Auth0)

**Reply by bamofah** [reply]
@balazsorban44 Question. When u say used your own custom provider do you mean you where able to make use of the oauth0 package put configure it (urls etc) to point to your custom oauth server??

**Reply by balazsorban44** [reply]
@bamofah sorry for the late response, yes! we have [IdentityServer 4](https://identityserver4.readthedocs.io/en/latest/), and it works well!

**Reply by bamofah** [reply]
@balazsorban44 thanks for the reply 
so in that case did you use the package nextjs-auth0 but pointed it to our IdentServer 4? asking as I got an spring auth-server so was trying to figure how best to handle two specific oauth-flows using next js.

**Reply by balazsorban44** [reply]
Yes, using IDS4 with @auth0/nextjs-auth0

**Reply by bamofah** [reply]
@balazsorban44  nice will have to try and see if I can get it to talk to my spring-server

**Comment by balazsorban44** [comment]
Used the example they have given here:
https://github.com/auth0/nextjs-auth0/blob/master/examples/basic-example/pages/api/me.js
to
create an API endpoint, to retrieve the user info with a single fetch call
to /api/me. It can nicely be wrapped with
https://github.com/auth0/nextjs-auth0/blob/master/examples/basic-example/lib/user.js
for
convenience. I made some modifications to the code, but the principle is
the same.

**Comment by irving-caamal** [comment]
https://github.com/nextauthjs highly recommended library which allows you to consume both popular(OAuth, google) and custom providers. with best OWASP practices

**Comment by vhakulinen** [comment]
One way you can do this is to create a layout which you use in all your admin pages and then in that layout do the permission checking.

**Reply by andremarcondesteixeira** [reply]
I would like to understand the "downvotes". This looks like a solution.

**Reply by guitheengineer** [reply]
Can be because this solution has the same problem below, which OP tried to solve:
> The downside is that you need to repeat this for every page. [...] Is there a way to do this globally for everything under this path
